admin/database/postgres/postgres.go

```diff
@@ -840,36 +781,126 @@ func (c *connection) FindProjectInvite(ctx context.Context, projectID, userEmail
+		return parseErr(target, err)
+	}
+	if n == 0 {
+		return parseErr(target, sql.ErrNoRows)
+	}
+	if n > 1 {
+		// This should never happen
+		panic(fmt.Errorf("expected to update 1 row, but updated %d", n))
+	}
+	return nil
+}
+
+func checkDeleteRow(target string, res sql.Result, err error) error {
 	if err != nil {
-		return parseErr(err)
+		return parseErr(target, err)
+	}
+	n, err := res.RowsAffected()
+	if err != nil {
+		return parseErr(target, err)
+	}
+	if n == 0 {
+		return parseErr(target, sql.ErrNoRows)
+	}
+	if n > 1 {
+		// This should never happen
+		panic(fmt.Errorf("expected to delete 1 row, but deleted %d", n))
 	}
 	return nil
 }
 
-func parseErr(err error) error {
+func parseErr(target string, err error) error {
 	if err == nil {
 		return nil
 	}
 	if errors.Is(err, sql.ErrNoRows) {
-		return database.ErrNotFound
+		if target == "" {
+			return database.ErrNotFound
+		}
+		return &wrappedError{
+			msg: fmt.Sprintf("%s not found", target),
+			// wrap database.ErrNotFound so checks with errors.Is(...) still work
+			err: database.ErrNotFound,
+		}
+	}
+	var pgerr *pgconn.PgError
+	if !errors.As(err, &pgerr) {
+		return err
 	}
-	if strings.Contains(strings.ToLower(err.Error()), "violates unique constraint") {
-		return database.ErrNotUnique
+	if pgerr.Code == "23505" { // unique_violation
+		switch pgerr.ConstraintName {
+		case "orgs_name_idx":
+			return newAlreadyExistsErr("an org with that name already exists")
+		case "projects_name_idx":
+			return newAlreadyExistsErr("a project with that name already exists in the org")
+		case "users_email_idx":
+			return newAlreadyExistsErr("a user with that email already exists")
+		case "usergroups_name_idx":
+			return newAlreadyExistsErr("a usergroup with that name already exists in the org")
+		case "usergroups_users_pkey":
+			return newAlreadyExistsErr("user is already a member of the usergroup")
+		case "users_orgs_roles_pkey":
+			return newAlreadyExistsErr("user is already a member of the org")
+		case "users_projects_roles_pkey":
+			return newAlreadyExistsErr("user is already a member of the project")
+		case "usergroups_orgs_roles_pkey":
+			return newAlreadyExistsErr("usergroup is already a member of the org")
+		case "usergroups_projects_roles_pkey":
+			return newAlreadyExistsErr("usergroup is already a member of the project")
+		case "org_invites_email_org_idx":
+			return newAlreadyExistsErr("email has already been invited to the org")
+		case "project_invites_email_project_idx":
+			return newAlreadyExistsErr("email has already been invited to the project")
+		default:
+			if target == "" {
+				return database.ErrNotUnique
+			}
+			return newAlreadyExistsErr(fmt.Sprintf("%s already exists", target))
+		}
 	}
 	return err
 }
+
+func newAlreadyExistsErr(msg string) error {
+	// wrap database.ErrNotUnique so checks with errors.Is(...) still work
+	return &wrappedError{msg: msg, err: database.ErrNotUnique}
+}
+
+type wrappedError struct {
+	msg string
```

Comment:
1. Error codes are not really typical unless communicating across serialization boundaries. Instead, you would use custom types or wrap error constants – chose the latter here as they're easier to check for (can just do `errors.Is(err, database.ErrNotFound` because it's wrapping that).
2. The database layer should not return gRPC errors. These functions can also be called from REST APIs, background workers, etc.

Reply:
1. I am not sure of the usual practice but one argument in favour of using codes instead of wrapped errors is that it would make switch statements easier(In cases when error has additional context and `errrors.Unwrap` is required). 
2. I meant at API layer only.